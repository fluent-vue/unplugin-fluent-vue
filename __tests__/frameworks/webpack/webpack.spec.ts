import { describe, expect, it } from 'vitest'

import type { Stats } from 'webpack'

import { compile } from './util'

describe('webpack loader', () => {
  it('generates custom block code', async () => {
    // Arrange
    // Act
    const stats = await compile('fixtures/test.vue')

    // Assert
    const ftlModules = stats.toJson({ source: true }).modules
      ?.filter(module => module.name?.includes('blockType=fluent') === true)
      .map(module => module.source)

    expect(ftlModules).not.toBeUndefined()
    expect(ftlModules).toHaveLength(1)
    expect(ftlModules).toMatchSnapshot()
  })

  it('includes HMR code only when HMR is enabled', async () => {
    // Arrange
    // Act
    const stats = await compile('fixtures/test.vue', {}, true)

    // Assert
    const ftlModules = stats.toJson({ source: true }).modules
      ?.filter(module => module.name?.includes('blockType=fluent') === true)
      .map(module => module.source)

    expect(ftlModules).not.toBeUndefined()
    expect(ftlModules).toHaveLength(1)
    expect(ftlModules).toMatchSnapshot()
  })

  it('errors with no locale attr', async () => {
    // Arrange
    const func = async (): Promise<Stats> => await compile('fixtures/noLocale.vue')

    // Act
    // Assert
    await expect(func).rejects.toContainEqual(
      expect.objectContaining({
        details: expect.stringContaining('Error: Custom block does not have locale attribute'),
      }))
  })
})
