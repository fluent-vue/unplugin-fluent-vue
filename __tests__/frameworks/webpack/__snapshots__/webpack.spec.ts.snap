// Vitest Snapshot v1

exports[`Webpack SFC > generates custom block code 1`] = `
[
  "export { default } from \\"-!../../node_modules/.pnpm/unplugin@0.8.0_vite@3.0.3+webpack@5.73.0/node_modules/unplugin/dist/webpack/loaders/transform.js??unplugin-fluent-vue-sfc!../../node_modules/.pnpm/vue-loader@17.0.0_webpack@5.73.0/node_modules/vue-loader/dist/index.js??ruleSet[1].rules[2].use[0]!./test.vue?vue&type=custom&index=0&blockType=fluent&locale=en\\"; export * from \\"-!../../node_modules/.pnpm/unplugin@0.8.0_vite@3.0.3+webpack@5.73.0/node_modules/unplugin/dist/webpack/loaders/transform.js??unplugin-fluent-vue-sfc!../../node_modules/.pnpm/vue-loader@17.0.0_webpack@5.73.0/node_modules/vue-loader/dist/index.js??ruleSet[1].rules[2].use[0]!./test.vue?vue&type=custom&index=0&blockType=fluent&locale=en\\"",
  "
import { FluentResource } from '@fluent/bundle'

export default function (Component) {
  const target = Component.options || Component
  target.fluent = target.fluent || {}
  target.fluent['en'] = new FluentResource(\`# Simple things are simple.
hello-user = Hello, {$userName}!

# Complex things are possible.
shared-photos =
  {$userName} {$photoCount ->
     [one] added one photo
    *[other] added {$photoCount} new photos
  } to {$userGender ->
     [male] his stream
     [female] her stream
    *[other] their stream
  }.\`)
}
",
]
`;

exports[`Webpack SFC > includes HMR code only when HMR is enabled 1`] = `
[
  "export { default } from \\"-!../../node_modules/.pnpm/unplugin@0.8.0_vite@3.0.3+webpack@5.73.0/node_modules/unplugin/dist/webpack/loaders/transform.js??unplugin-fluent-vue-sfc!../../node_modules/.pnpm/vue-loader@17.0.0_webpack@5.73.0/node_modules/vue-loader/dist/index.js??ruleSet[1].rules[2].use[0]!./test.vue?vue&type=custom&index=0&blockType=fluent&locale=en\\"; export * from \\"-!../../node_modules/.pnpm/unplugin@0.8.0_vite@3.0.3+webpack@5.73.0/node_modules/unplugin/dist/webpack/loaders/transform.js??unplugin-fluent-vue-sfc!../../node_modules/.pnpm/vue-loader@17.0.0_webpack@5.73.0/node_modules/vue-loader/dist/index.js??ruleSet[1].rules[2].use[0]!./test.vue?vue&type=custom&index=0&blockType=fluent&locale=en\\"",
  "
import { FluentResource } from '@fluent/bundle'

export default function (Component) {
  const target = Component.options || Component
  target.fluent = target.fluent || {}
  target.fluent['en'] = new FluentResource(\`# Simple things are simple.
hello-user = Hello, {$userName}!

# Complex things are possible.
shared-photos =
  {$userName} {$photoCount ->
     [one] added one photo
    *[other] added {$photoCount} new photos
  } to {$userGender ->
     [male] his stream
     [female] her stream
    *[other] their stream
  }.\`)
}
",
]
`;
